orchestra/static/orchestra/task/js/autosave.js

```diff
@@ -0,0 +1,87 @@
+(function() {
+  'use strict';
+
+  var serviceModule =  angular.module('orchestra.task.services');
+
+  serviceModule.factory('autoSaveTask', function($rootScope, $timeout, $http, orchestraService) {
+    var autoSaveTask = new function() {
+      var autoSaver = this;
+
+      autoSaver.setup = function($scope, taskId, taskData) {
+        autoSaver.saveError = false;
+        autoSaver.saving = false;
+        autoSaver.timeout = 10000;
+        autoSaver.taskId = taskId;
+        autoSaver.data = taskData;
+
+        var handler = $rootScope.$on('task.data:change', function() {
+          autoSaver.schedule();
+        });
+        $scope.$on('$destroy', handler);
+
+        // Browser close or reload
+        window.onbeforeunload = function() {
+          if (autoSaver.autoSaveTimer || autoSaver.saveError) {
+            return 'Your latest changes haven\'t been saved.'
+          }
+        }
+
+        // Angular location change
+        $scope.$on('$locationChangeStart', function(e) {
+          if (autoSaver.autoSaveTimer || autoSaver.saveError) {
+            if (!confirm('Your latest changes haven\'t been saved.\n\n' +
+                  'Are you sure you want to leave this page?')) {
+              e.preventDefault();
+            }
+          }
+        });
+      }
+
+      autoSaver.schedule = function() {
+        if (!autoSaver.autoSaveTimer && !autoSaver.is_read_only) {
```

Comment:
likely from the scope and not from autosaver?

let's manually test
- autosave works
- save works
- no autosave on readonly tasks